README.md
# README

将数据从缓存持久化到数据库

---

## Table of Contents

<!-- vim-markdown-toc GFM -->

* [DataWizard](#datawizard)
* [配置文件](#配置文件)

<!-- vim-markdown-toc -->

---

[PostgreSQL性能优化](http://mysql.taobao.org/monthly/2016/04/05/)

[PostgreSQL写入性能优化](https://developer.aliyun.com/article/647444)

[PostgreSQL批量更新/插入](https://www.jianshu.com/p/1e389047cfa6)

[PostgreSQL 如何实现批量更新、删除、插入](https://developer.aliyun.com/article/74420)

[python使用psycopg2批量插入数据](https://blog.csdn.net/lsr40/article/details/83537974)

[psycopg2批量插入方法对比](https://blog.csdn.net/china1987427/article/details/95120023)

---

## DataWizard

1. 按设备从MQTT订阅数据

2. 使用批量INSERT

3. 有两种数据分批方案：

    1. 按时间分片

        主要可能出现数采频率不高导致一定时间片（例如1s，再长可能影响数据实时性）内只有一条数据，结果还是单条INSERT

    2. 按队列大小

        如果queue.qsize()到了一定大小，则采集这批数据，但可能出现达到指定大小耗时过长的问题

## 配置文件

- [main]部分：

    1. `number`是worker*基数*，它的计算公式为：

        数据源（假设是mqtt）有**10**个topic，每个topic发布的数据频率F=**100**Hz，服务器持久化性能P=0.01s/条，系数C=10，则`number`计算公式如下：

        ```shell
        number = F * P * C
        ```

        得出number = 10

        > 系数C由服务器性能、网络环境等因素综合得出，不是一个定值


CHANGELOG.md
# CHANGELOG

CHANGELOG

---

## Table of Contents

<!-- vim-markdown-toc GFM -->

* [v0.8.4](#v084)
* [v0.8.3](#v083)
* [v0.8.2](#v082)
* [v0.8.1](#v081)
* [v0.8.0](#v080)
* [v0.7.3](#v073)
* [v0.7.2](#v072)
* [v0.7.1](#v071)
* [v0.7.0](#v070)
* [v0.6.2](#v062)
* [v0.6.1](#v061)
* [v0.6.0](#v060)
* [v0.5.2](#v052)
* [v0.4.0](#v040)
* [v0.3.0](#v030)
* [v0.2.0](#v020)
* [v0.1.2](#v012)
* [v0.1.1](#v011)
* [v0.1](#v01)

<!-- vim-markdown-toc -->

---

更新日志

---

## v0.8.4

> 2021-11-03 13:47

- rename:
    - test --> scripts
    - test/mqtt --> test/MQTT
    - test/timescaledb --> test/PostgreSQL

## v0.8.3

> 2021-09-23 15:57

- 提示clientid不能重复问题
- 调整日志参数
- 修改compose文件不使用PWD变量
- 从queue.Queue切换回multiprocessing.Queue: 多线程和多进程通信队列的区别

## v0.8.2

> 2021-07-28 11:39

- 线程数减少问题已排除，现在使用新版入库函数
- 当配置了worker数时不再限制其最小值
- 将测试用的多线程方法合并后删除测试用例
- 修改插件防止其返回值可能是None

## v0.8.1

> 2021-06-23 16:31

- 添加数据结构检查函数，通过检查的数据才会进入入库流程，不通过的则被放弃

## v0.8.0

> 2021-06-02 15:32

- start_source和start_wizard分到两个进程运行
- 重写mqtt_wrapper，函数取代类

## v0.7.3

> 2021-06-01 17:10

- mqtt_wrapper修改前最后一个tag

## v0.7.2

> 2021-05-26 18:29

- 添加MQTT重连方法

## v0.7.1

> 2021-04-27 10:30

- 补充log
- fixed bug

## v0.7.0

> 2021-04-26 17:26

- `insert_nextgen`全面替换`insert`方法

## v0.6.2

> 2021-04-26 15:10

- 调整worker数，并设置最小为CPU核心数+4
- 使worker数和tasks数保持一致

## v0.6.1

> 2021-04-25 16:44

- 调整配置文件结构

## v0.6.0

> 2021-04-25 13:23

- 全面从insert切换到insert_nextgen

## v0.5.2

> 2021-04-21 08:17

- 添加版本号
- 增加对json数据的支持
- 加入插件系统
- 增加CHANGELOG

## v0.4.0

> 2021-01-13 15:11

- 更新数据库相关配置

## v0.3.0

> 2021-01-05 09:55

- 修复一些小错误
- 准备改版

## v0.2.0

> 2020-12-29 16:45

- 重构mqtt操作代码

## v0.1.2

> 2020-12-29 11:58

- 提供compose文件

## v0.1.1

> 2020-12-29 10:42

- 修复log记录问题

## v0.1

> 2020-12-28 18:08

- 初步完成


main_dev.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
File: main.py
Author: YJ
Email: yj1516268@outlook.com
Created Time: 2020-11-06 16:06:47

Description: 将data和log从缓存(redis, mqtt ...)持久化到数据库(TimescaleDB)

"""

import json
import logging
import os
import time
from concurrent.futures import ThreadPoolExecutor
from queue import Queue

import toml

from utils.log_wrapper import setup_logging
from utils.mqtt_wrapper import MqttWrapper
from utils.timescale_wrapper_forklog import TimescaleWrapper

logger = logging.getLogger('DataWizard.main')


class Wizard(object):
    """Data Wizard."""
    def __init__(self, conf):
        """Initialize.

        :conf: 总配置信息

        """
        self.conf = conf

        # [main] - Dizard配置
        main_conf = conf.get('main', dict())
        # # 进程或线程数
        self.number = main_conf.get('number') + os.cpu_count(
        ) if main_conf.get('number', 0) > 0 else os.cpu_count()
        data_source = main_conf.get('data_source', 'mqtt')
        data_storage = main_conf.get('data_storage', 'timescaledb')

        # 主要配置部分
        source_conf = conf['source'].get(data_source, dict())
        storage_conf = conf['storage'].get(data_storage, dict())

        # 根据topic数量动态构造数据缓存队列的字典
        self.topics = source_conf.get('topics', list())
        queues = [Queue() for _ in range(len(self.topics))]
        self.queue_dict = dict(zip(self.topics, queues))

        # [source] - 数据来源配置
        if data_source == 'mqtt':
            self.mqtt = MqttWrapper(source_conf, self.queue_dict)

        # [storage] - 数据去处配置
        if data_storage == 'timescaledb':
            self.database = TimescaleWrapper(storage_conf)

        # [log] - Log记录器配置
        setup_logging(conf['log'])

    def convert(self, raw_data):
        """Convert data
        :returns: data

        """
        data_str = raw_data.decode('UTF-8')
        data = json.loads(data_str)

        return data

    def persistence(self, topic):
        """数据持久化

        :topic: topic name

        """
        while True:
            queue = self.queue_dict.get(topic)
            data_bytes = queue.get()
            qsize = queue.qsize()

            data = self.convert(data_bytes)
            # TODO: 调用数据解析函数 <31-12-20, YJ> #
            _start = time.time()
            self.database.insert(data)
            _end = time.time()
            logger.info(
                ("Got the data, "
                 "Queue ({name}) size = {size} "
                 "<--> Time cost = {cost}s").format(name=topic,
                                                    size=qsize,
                                                    cost=_end - _start))
            if qsize >= 5000:
                break

    def start_mqtt(self):
        """启动Mqtt客户端订阅数据"""
        self.mqtt.sub_message()

    def start_wizard(self):
        """Main."""
        with ThreadPoolExecutor(max_workers=self.number * len(self.topics),
                                thread_name_prefix='Wizard') as executor:
            executor.map(self.persistence,
                         self.topics * self.number,
                         chunksize=len(self.topics))


if __name__ == "__main__":
    logger.info('Action')

    confile = './conf/conf.toml'
    conf = toml.load(confile)
    wizard = Wizard(conf)

    wizard.start_mqtt()
    wizard.start_wizard()
